parlai/tasks/flattened/agents.py

```diff
@@ -0,0 +1,232 @@
+            '--invalidate-cache',
+            type='bool',
+            default=False,
+            help='Set this to True to rebuild the data',
+        )
+
+        # Add the arguments for the teacher
+        opt = parser.parse_and_process_known_args()[0]
+        tasks = get_original_task(opt, multi_possible=True)
+        for task in tasks:
+            if hasattr(task, 'add_cmdline_args'):
+                task.add_cmdline_args(parser)
+
+        return parser
+
+    def __init__(self, opt, shared=None):
+        self.opt = opt
+
+        if shared and 'data' in shared:
+            self.data = shared['data']
+        else:
+            self.data = self._setup_data(opt)
+
+        super().__init__(opt, shared)
+        self.reset()
+
+    def num_episodes(self):
+        return len(self.data)
+
+    def num_examples(self):
+        return len(self.data)
+
+    def _setup_data(self, opt, context_length=-1):
+        # possibly make new data directory
+        original_task = get_original_task(opt)
+        self.original_task = ':'.join(opt['task'].split(':')[2:])
+        teacher_name = self.original_task.replace(':', '_') + '_flattened'
+        self.save_dir = os.path.join(opt['datapath'], teacher_name)
+        os.makedirs(self.save_dir, exist_ok=True)
+
+        datatype = opt['datatype'].split(':')[0]
+        self.save_path = os.path.join(self.save_dir, f'{datatype}_data.json')
+
+        data = self.load_data(opt)
+        if data is not None:
+            # successfully load data
+            return data
+
+        # build the original teacher
+        teacher_opt = deepcopy(opt)
+        teacher_opt['task'] = self.original_task
+        teacher = original_task(teacher_opt)
+
+        # did not load data, let's build it!
+        total_exs = teacher.num_examples()
+        num_exs = 0
+
+        context_length = opt['max_context_length']
+        include_labels = opt['flatten_include_labels']
+        delimiter = opt['flatten_delimiter']
+
+        total_eps = teacher.num_episodes()
+        progress_bar = tqdm(
+            total=total_eps, unit='ex', unit_scale=True, desc='Building flattened data'
+        )
+
+        current_episode = []
+        episode_done = False
+
+        all_episodes = []
+        while num_exs < total_exs:
+            while not episode_done:
+                # TODO: eventually all teachers should return Messages, so
+                # we should assert this
+                action = Message(teacher.act())
+                current_episode.append(action)
+                episode_done = action.get('episode_done', False)
+                num_exs += 1
+
+            # flatten the episode into 1-example episodes with context
+            flattened_ep = flatten(
+                current_episode,
+                context_length,
+                include_labels=include_labels,
+                delimiter=delimiter,
+            )
+            all_episodes += flattened_ep
+
+            current_episode = []
```

Comment:
nit: you can reset these two variables above the `while not episode_done` statement to avoid having to write it twice